### PR TITLE
feat(core): add utility methods to NameOrAddress

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

@gakonst https://t.me/ethers_rs/20599

## Solution

- fix `impl Encodable for NameOrAddress` wrong cmp order
- add as_X -> X by cloning to NameOrAddress
- add to_addr to TransactionRequest
- add more impls and cleanup

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Updated the changelog

### CODE REVIEW

File: ethers-core/src/types/ens.rs [hunk 1]
```diff
@@ -85,8 +72,57 @@ impl<'de> Deserialize<'de> for NameOrAddress {
...
+    /// Maps Address(&a) to Some(a) by cloning the value and Name to None.
+    pub fn as_address(&self) -> Option<Address> {
+        match self {
+            Self::Address(a) => Some(*a),
+            Self::Name(_) => None,
+        }
+    }
+
+    /// Maps Name(&n) to Some(n) by cloning the value and Address to None.
+    pub fn as_name(&self) -> Option<String> {
```
Comment (mattsse) [inline ethers-core/src/types/ens.rs:121]: this should return `&str` so caller can decided whether to clone

**Review by mattsse — CHANGES_REQUESTED** [review]
smol nit, otherwise lgtm

File: ethers-core/src/types/ens.rs (hunk 1 above)
Comment (DaniPopes) [reply ethers-core/src/types/ens.rs:121]: Ok, same for `as_address`?

Comment (mattsse) [reply ethers-core/src/types/ens.rs:121]: `Address` is `Copy` so nbd there, but for consistency maybe it makes sense

**Review by gakonst — APPROVED** [review]
beautiful - clean

**Pull request merged**